acceptance/helpers.rb

```diff
@@ -66,173 +203,205 @@ def prepare_upgrade_with(initial_package_version_or_collection)
+          teardowns__clean_agents.each do |teardown__clean_agents|
+            teardown(&teardown__clean_agents)
           end
         end
       end
 
-      # Wraps {Beaker::DSL::PuppetHelpers.with_puppet_running_on} to apply a default
-      # manifest to all nodes and execute a block. Behaves as follows:
+      # Applies a manifest on the agents. Behaves as follows:
       #
       # 1. Set up a `site.pp` file in the production environment on the master
-      #    such that the puppet code in `site_pp_contents` is applied to all nodes.
+      #    such that the puppet code in `manifest` is applied _only_ on the agents.
       #    - If a `site.pp` already exists there, record its contents and
-      #      permissions and create a teardown task to restore them after the test
-      #      finishes.
-      # 2. Using {Beaker::DSL::PuppetHelpers.with_puppet_running_on}:
-      #    - Perform a puppet run on the agents, and
-      #    - execute the block, if any.
-      #
-      # @see Beaker::DSL::PuppetHelpers.with_puppet_running_on
-      # @param [String] site_pp_contents The contents of the default site.pp file. This
-      #   content will be wrapped as follows, so that it applies to all nodes:
+      #      permissions, then restore them after the method's invoked.
+      #
+      # 2. Perform a puppet run on the agents
+      #
+      # @param [String] manifest The manifest to apply. This content will
+      #   be wrapped as follows, so that it applies to all of the agents:
       #     ```
-      #     node default {
-      #       #{site_pp_contents}
+      #     node '#{agent1}', '#{agent2}' ... {
+      #       #{manifest}
       #     }
-      #     ```
-      # @param [Hash] master_opts Options to pass to {Beaker::DSL::PuppetHelpers.with_puppet_running_on}.
-      # @yield Invokes {Beaker::DSL::PuppetHelpers.with_puppet_running_on},
-      # passing along master_opts, if supplied.
-      def with_default_site_pp(site_pp_contents, master_opts = {})
-        manifest_contents = %(node default { #{site_pp_contents} })
+      def apply_manifest_on_agents(manifest)
+        agent_nodes = agents_only.map do |agent|
+          "'#{agent.to_s}'"
+        end.join(', ')
+        site_pp_contents = %(node #{agent_nodes} { #{manifest} })
 
         # PMT will have installed dependencies in the production environment; We will put our manifest there, too:
         site_pp_path = File.join(puppet_config(master, 'codedir'), 'environments', 'production', 'manifests', 'site.pp')
 
-        step "Save current site.pp" do
+        cleanup = nil
+
+        step "(Master) Save the current site.pp file" do
           if file_exists_on(master, site_pp_path)
             original_contents = file_contents_on(master, site_pp_path)
             original_perms = on(master, %(stat -c "%a" #{site_pp_path})).stdout.strip
-
-            teardown do
-              step "restore original manifest" do
+  
+            cleanup = lambda do
```

Comment:
What was the reason to change the teardowns to lambdas?

Reply:
This is so we'd reset the `site.pp` file after each method invocation instead of after the test finishes. That way, this function can be re-used many times in the same test.